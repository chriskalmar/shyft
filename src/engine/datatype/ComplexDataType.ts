export class ComplexDataType {}

export const isComplexDataType = (obj: unknown): obj is ComplexDataType => {
  return obj instanceof ComplexDataType;
};
