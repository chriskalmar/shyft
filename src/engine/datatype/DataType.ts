import { Entity, ShadowEntity, ViewEntity } from '../..';
import { passOrThrow, isFunction } from '../util';
import { ComplexDataType } from './ComplexDataType';

export type DataTypeSetup = {
  name?: string;
  description?: string;
  mock?: () => any;
  validate?: () => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;
};

export type DataTypeFunction = (params: {
  setup: DataTypeSetup;
  entity?: Entity | ShadowEntity | ViewEntity;
}) => DataType | ComplexDataType;

export class DataType {
  name: string;
  description: string;
  mock?: () => any;
  validator?: (value: any, context: any) => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;

  constructor(setup: DataTypeSetup = {} as DataTypeSetup) {
    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup;

    passOrThrow(name, () => 'Missing data type name');
    passOrThrow(
      description,
      () => `Missing description for data type '${name}'`,
    );

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`,
    );

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`,
      );

      this.validator = validate;
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`,
      );

      this.defaultValue = defaultValue;
    }

    this.name = name;
    this.description = description;
    this.mock = mock;

    if (enforceRequired) {
      this.enforceRequired = enforceRequired;
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex;
    }
  }

  validate = async (value: any, context: any): Promise<void> => {
    if (value && this.validator) {
      await this.validator(value, context);
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isDataType = (obj: unknown): obj is DataType => {
  return obj instanceof DataType;
};
