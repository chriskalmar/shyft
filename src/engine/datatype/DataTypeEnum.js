import { passOrThrow, isArray } from '../util';

import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';

import DataType from './DataType';

export class DataTypeEnum extends DataType {
  constructor(setup = {}) {
    const { name, description, values } = setup;

    passOrThrow(
      isArray(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    values.map(value => {
      passOrThrow(
        enumValueRegex.test(value),
        () =>
          `Invalid enum value '${value}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );
    });

    super({
      ...setup,
      description: description || `Enumeration set: ${values.join(', ')}`,
      mock() {
        const randomPos = Math.floor(Math.random() * values.length);
        return values[randomPos];
      },
    });

    this.values = values;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeEnum = obj => {
  return obj instanceof DataTypeEnum;
};
