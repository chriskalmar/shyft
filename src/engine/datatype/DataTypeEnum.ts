import { uniq } from 'lodash';
import { passOrThrow, isMap } from '../util';
import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';
import { DataType } from './DataType';

export type ValueType = {
  [key: string]: number;
};

export type DataTypeEnumSetupType = {
  name: string;
  description?: string;
  values: ValueType;
};

export class DataTypeEnum extends DataType {
  values: ValueType;

  constructor(setup: DataTypeEnumSetupType = {} as DataTypeEnumSetupType) {
    const { name, description, values } = setup;

    passOrThrow(
      isMap(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    const valueNames = Object.keys(values);
    const uniqueIds = [];

    valueNames.map((valueName) => {
      const valueId = values[valueName];
      uniqueIds.push(valueId);

      passOrThrow(
        enumValueRegex.test(valueName),
        () =>
          `Invalid value name '${valueName}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );

      passOrThrow(
        valueId === Number(valueId) && valueId > 0,
        () =>
          `Value '${valueName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === uniq(uniqueIds).length,
      () =>
        `Each value defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      name,
      description: description || `Enumeration set: ${valueNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.values = values;
  }

  toString(): string {
    return this.name;
  }
}

export const isDataTypeEnum = (obj: any): boolean => {
  return obj instanceof DataTypeEnum;
};
