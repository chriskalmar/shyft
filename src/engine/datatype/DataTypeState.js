import { passOrThrow, isMap } from '../util';

import { stateNameRegex, STATE_NAME_PATTERN } from '../constants';

import * as _ from 'lodash';

import DataType from './DataType';

export class DataTypeState extends DataType {
  constructor(setup = {}) {
    const { name, description, states } = setup;

    passOrThrow(
      isMap(states, true),
      () => `'Missing states for data type '${name}'`,
    );

    const stateNames = Object.keys(states);
    const uniqueIds = [];

    stateNames.map(stateName => {
      const stateId = states[stateName];
      uniqueIds.push(stateId);

      passOrThrow(
        stateNameRegex.test(stateName),
        () =>
          `Invalid state name '${stateName}' for data type '${name}' (Regex: /${STATE_NAME_PATTERN}/)`,
      );

      passOrThrow(
        stateId === parseInt(stateId, 10) && stateId > 0,
        () =>
          `State '${stateName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === _.uniq(uniqueIds).length,
      () =>
        `Each state defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      ...setup,
      description: description || `States: ${stateNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.states = states;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeState = obj => {
  return obj instanceof DataTypeState;
};
