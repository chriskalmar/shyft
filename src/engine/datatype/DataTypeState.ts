import { passOrThrow, isMap } from '../util';

import { stateNameRegex, STATE_NAME_PATTERN } from '../constants';

import * as _ from 'lodash';

import { DataType } from './DataType';

export type StateType = {
  [key: string]: number;
};

export type DataTypeStateSetupType = {
  name: string;
  description?: string;
  states: StateType;
};

export class DataTypeState extends DataType {
  states: StateType;
  constructor(setup: DataTypeStateSetupType = <DataTypeStateSetupType>{}) {
    const { name, description, states } = setup;

    passOrThrow(
      isMap(states, true),
      () => `'Missing states for data type '${name}'`,
    );

    const stateNames = Object.keys(states);
    const uniqueIds = [];

    stateNames.map(stateName => {
      const stateId = states[stateName];
      uniqueIds.push(stateId);

      passOrThrow(
        stateNameRegex.test(stateName),
        () =>
          `Invalid state name '${stateName}' for data type '${name}' (Regex: /${STATE_NAME_PATTERN}/)`,
      );

      passOrThrow(
        stateId === Number(stateId) && stateId > 0,
        () =>
          `State '${stateName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === _.uniq(uniqueIds).length,
      () =>
        `Each state defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      name,
      description: description || `States: ${stateNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.states = states;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeState = obj => {
  return obj instanceof DataTypeState;
};
