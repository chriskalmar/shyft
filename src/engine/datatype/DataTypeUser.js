import DataType from './DataType';

export class DataTypeUser extends DataType {}

export const isDataTypeUser = obj => {
  return obj instanceof DataTypeUser;
};
