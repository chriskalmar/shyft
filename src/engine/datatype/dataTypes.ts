import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
