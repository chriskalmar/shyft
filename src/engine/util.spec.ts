/* eslint-disable @typescript-eslint/explicit-function-return-type */

import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isArray,
  mergeMaps,
  mapOverProperties,
  sortDataByKeys,
  reverseString,
} from './util';

import { DataTypeBoolean } from './datatype/dataTypes';

describe('util', () => {
  describe('passOrThrow', () => {
    it('should pass on positive condition', () => {
      function fn() {
        passOrThrow(1 < 2, () => 'This error will never happen');
      }

      expect(fn).not.toThrow();
    });

    it('should throw on negative condition', () => {
      function fn() {
        passOrThrow(1 > 2, () => 'This is the very very super important error');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if message is not a function', () => {
      function fn() {
        passOrThrow(1 > 0, 'This is the very very super important error');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('resolveFunctionMap', () => {
    const dataMap = {
      a: 123,
      b: 456,
    };

    it('should return the provided map', () => {
      const result = resolveFunctionMap(dataMap);

      expect(result).toEqual(dataMap);
    });

    it('should return the result of the provided function', () => {
      const result = resolveFunctionMap(() => {
        return dataMap;
      });

      expect(result).toEqual(dataMap);
    });
  });

  describe('isMap', () => {
    it('should accept maps', () => {
      expect(isMap({})).toBe(true);
      expect(isMap({ a: 123 })).toBe(true);
      expect(isMap(Object.create({}))).toBe(true);
      expect(isMap(DataTypeBoolean)).toBe(true);
    });

    it('should reject non-maps', () => {
      expect(isMap()).toBe(false);
      expect(isMap(null)).toBe(false);
      expect(isMap(undefined)).toBe(false);
      expect(isMap([])).toBe(false);
      expect(isMap([1, 2, 3])).toBe(false);
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      expect(isMap(() => {})).toBe(false);
      expect(isMap(1234567)).toBe(false);
    });

    it('should reject empty maps if flag `nonEmpty` is set', () => {
      expect(isMap({}, true)).toBe(false);
    });
  });

  describe('isArray', () => {
    it('should accept arrays', () => {
      expect(isArray([])).toBe(true);
      expect(isArray(['test'])).toBe(true);
      expect(isArray([1, 2, 3])).toBe(true);
    });

    it('should reject non-arrays', () => {
      expect(isArray()).toBe(false);
      expect(isArray(null)).toBe(false);
      expect(isArray(undefined)).toBe(false);
      expect(isArray(Object.create({}))).toBe(false);
      expect(isArray(DataTypeBoolean)).toBe(false);
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      expect(isArray(() => {})).toBe(false);
      expect(isArray(1234567)).toBe(false);
    });

    it('should reject empty arrays if flag `nonEmpty` is set', () => {
      expect(isArray([], true)).toBe(false);
    });
  });

  describe('mergeMaps', () => {
    it('should merge 2 maps', () => {
      let obj;

      obj = mergeMaps({}, {});
      expect(obj).toEqual({});

      obj = mergeMaps({ a: 123 }, {});
      expect(obj).toEqual({ a: 123 });

      obj = mergeMaps({}, { b: 456 });
      expect(obj).toEqual({ b: 456 });

      obj = mergeMaps({ a: 123 }, { b: 456 });
      expect(obj).toEqual({ a: 123, b: 456 });

      obj = mergeMaps({ b: 456 }, { a: 123 });
      expect(obj).toEqual({ a: 123, b: 456 });

      obj = mergeMaps({ a: 123, b: 456 }, { b: 789 });
      expect(obj).toEqual({ a: 123, b: 789 });

      obj = mergeMaps(
        { a: 123, b: [1, 2, 4], c: { deep: [7, 8, 9] } },
        { b: { b1: 1, b2: 2 } },
      );
      expect(obj).toEqual({
        a: 123,
        b: { b1: 1, b2: 2 },
        c: { deep: [7, 8, 9] },
      });
    });

    it('should throw if non-maps are provided', () => {
      function fn1() {
        mergeMaps();
      }

      function fn2() {
        mergeMaps({ a: 123 }, []);
      }

      function fn3() {
        mergeMaps('string', {});
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });
  });

  describe('mapOverProperties', () => {
    it('should map over properties and call interatee', () => {
      const keys = [];
      let sum = 0;

      function iteratee(val, key) {
        sum += val;
        keys.push(key);
      }

      const someMap = {
        a: 1,
        b: 4,
        c: 8,
      };

      mapOverProperties(someMap, iteratee);

      expect(keys).toEqual(['a', 'b', 'c']);
      expect(sum).toBe(13);
    });

    it('should throw if non-maps are provided', () => {
      function fn1() {
        mapOverProperties();
      }

      function fn2() {
        mapOverProperties([]);
      }

      function fn3() {
        mapOverProperties('string');
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });

    it('should throw if iteratee is not a function', () => {
      function fn1() {
        mapOverProperties({});
      }

      function fn2() {
        mapOverProperties({}, []);
      }

      function fn3() {
        mapOverProperties({}, 'string');
      }

      expect(fn1).toThrowErrorMatchingSnapshot();
      expect(fn2).toThrowErrorMatchingSnapshot();
      expect(fn3).toThrowErrorMatchingSnapshot();
    });
  });

  describe('sortDataByKeys', () => {
    it('should return empty result if keys list is empty or invalid', () => {
      const result1 = sortDataByKeys([], { a: 1, b: 3 });
      const result2 = sortDataByKeys({}, { a: 1, b: 3 });
      const result3 = sortDataByKeys(null, { a: 1, b: 3 });

      expect(result1).toEqual([]);
      expect(result2).toEqual([]);
      expect(result3).toEqual([]);
    });

    it('should return null-filled array if data list is empty or invalid', () => {
      const result1 = sortDataByKeys(['a', 'b'], null);
      const result2 = sortDataByKeys(['a'], {});
      const result3 = sortDataByKeys(['a', 'b', 'f'], []);

      expect(result1).toEqual([null, null]);
      expect(result2).toEqual([null]);
      expect(result3).toEqual([null, null, null]);
    });

    it('should sort data based on keys', () => {
      const data = [
        { id: 'a', val: 'lorem' },
        { id: 'b', val: 'ipsum' },
        { id: 'c', val: 'dolor' },
        { id: 'd', val: 'et' },
        { id: 'e', val: 'unde' },
        { id: 'f', val: 'omnis' },
        { id: 'g', val: 'iste' },
        { id: 'a', val: 'natus' },
        { id: 'b', val: 'voluptatem' },
        { id: 'c', val: 'doloremque' },
      ];

      const keys1 = ['a', 'c', 'f'];
      const keys2 = ['e', 'e', 'b', 'b', 'c'];
      const keys3 = ['g'];
      const keys4 = [
        'a-unknown',
        'f',
        'c-unknown',
        'g',
        'c',
        'd-unkown',
        'c',
        'hh',
      ];

      const result1 = sortDataByKeys(keys1, data);
      const result2 = sortDataByKeys(keys2, data);
      const result3 = sortDataByKeys(keys3, data);
      const result4 = sortDataByKeys(keys4, data);

      expect(result1).toEqual([
        { id: 'a', val: 'lorem' },
        { id: 'c', val: 'dolor' },
        { id: 'f', val: 'omnis' },
      ]);

      expect(result2).toEqual([
        { id: 'e', val: 'unde' },
        { id: 'e', val: 'unde' },
        { id: 'b', val: 'ipsum' },
        { id: 'b', val: 'voluptatem' },
        { id: 'c', val: 'dolor' },
      ]);

      expect(result3).toEqual([{ id: 'g', val: 'iste' }]);

      expect(result4).toEqual([
        null,
        { id: 'f', val: 'omnis' },
        null,
        { id: 'g', val: 'iste' },
        { id: 'c', val: 'dolor' },
        null,
        { id: 'c', val: 'doloremque' },
        null,
      ]);
    });
  });

  describe('reverseString', () => {
    it('should reverse a string', () => {
      expect(reverseString('hello')).toEqual('olleh');
      expect(reverseString('')).toEqual('');
      expect(reverseString('a')).toEqual('a');
      expect(reverseString('aBC')).toEqual('CBa');
      expect(reverseString(' x y ')).toEqual(' y x ');
    });
  });
});
