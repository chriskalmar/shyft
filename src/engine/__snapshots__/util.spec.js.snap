// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util mapOverProperties should throw if iteratee is not a function 1`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if iteratee is not a function 2`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if iteratee is not a function 3`] = `"Provided iteratee is not a function"`;

exports[`util mapOverProperties should throw if non-maps are provided 1`] = `"Provided object is not a map"`;

exports[`util mapOverProperties should throw if non-maps are provided 2`] = `"Provided object is not a map"`;

exports[`util mapOverProperties should throw if non-maps are provided 3`] = `"Provided object is not a map"`;

exports[`util mergeMaps should throw if non-maps are provided 1`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util mergeMaps should throw if non-maps are provided 2`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util mergeMaps should throw if non-maps are provided 3`] = `"mergeMaps() expects 2 maps for a merge to work"`;

exports[`util passOrThrow should throw if message is not a function 1`] = `"passOrThrow() expects messageFn to be a function"`;

exports[`util passOrThrow should throw on negative condition 1`] = `"This is the very very super important error"`;
